### DISCUSSION TITLE
Power lost due to hull resistance vs. thrust power output of engine is inconsistent

### DISCUSSION THREAD

**Discussion by mygamingaccount** [FakeFishGames/Barotrauma#12330] [Ideas and suggestions]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

Power is force times velocity.

$$P = V \cdot F$$

Force due to hull resistance scales with the square of the speed. Barotrauma calculates the resistance force correctly and scales it with the displacement of the sub:

https://github.com/Regalis11/Barotrauma/blob/c67f6688fdf4dcffa224711a0d4f4181d9a1fad4/Barotrauma/BarotraumaShared/SharedSource/Map/SubmarineBody.cs#L438-L439

$$R \propto V^2 \cdot m$$

If the hull resistance scales with the square of the speed, then we can see that power lost due to hull resistance scales with the cube of the speed.

$$P_{hull} = V \cdot R$$

$$P_{hull} \propto V^3 \cdot m$$

However, engines in Barotrauma output a thrust force proportional with their power input.

https://github.com/Regalis11/Barotrauma/blob/c67f6688fdf4dcffa224711a0d4f4181d9a1fad4/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Engine.cs#L176-L187

$$P_{eng} \propto F_{eng} $$

This creates a paradox: when the forces acting on the sub are equal, then so should the power be equal. But expressing speed via the resistance and the power loss due to drag shows a different story:

$$ V \propto \sqrt[3]{\frac{P_{hull}}{m}} \propto \sqrt[2]{\frac{R}{m}}$$

For $P_{hull}$ to be equal to $P_{eng}$ at the same time as $R$ is equal to $F_{eng}$, it is necessary to raise $F_{eng}$ to the power of $\frac{3}{2}$

Therefore instead of 

$$P_{eng} \propto F_{eng} $$

it should be

$$P_{eng} \propto {F_{eng}}^{1.5} $$

### Reproduction steps

Suggested change in 

https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Engine.cs

Change Line 183:
```cs
            currPowerConsumption = Math.Abs(targetForce) / 100.0f * powerConsumption;
```
to
```cs
             currPowerConsumption = (float)Math.Pow(Math.Abs(targetForce) / 100.0f, 1.5f) * powerConsumption;
```

### Bug prevalence

Happens every time I play

### Version

v1.0.21.0

### Which operating system did you encounter this bug on?

Linux

**Comment by mygamingaccount** [comment]
While testing I've noticed that an interesting action arises from overvolting: if 100% thrust is achieved at 100% power, having 50% power at 63% thrust instead at max reactor output power, therefore getting 200% overvoltage thus 200% thrust, you get 126% total thrust output :smiley:

Adjusting the overvoltage factor to follow a $V^{2/3}$ curve is trivial. Letting this small exploit in is interesting too :grin:

**Comment by Regalis11** [comment]
Thank you for the report!

I have to admit I don't understand physics or the formulas here well enough to fully grasp the issue. This part in particular is puzzling to me:

> This creates a paradox: when the forces acting on the sub are equal, then so should the power be equal. But expressing speed via the resistance and the power loss due to drag shows a different story:

$$ V \propto \sqrt[3]{\frac{P_{hull}}{m}} \propto \sqrt[2]{\frac{R}{m}}$$

Why is it necessary the case that the power should be equal if the forces are equal? Couldn't the relation of the power consumption and force of an engine be practically anything, for example some kind of a complex curve - let's say the force being almost nothing unless some minimum power threshold is reached, and behaving optimally at some specific range and very poorly on some other range?

**Comment by mygamingaccount** [comment]
It isn't **necessary** that Power be proportional with $F^{3/2}$ - however it is the simplest solution to provide a flat efficiency curve, where **Power output = Power input *times an arbitrary number*** (since propulsive power is impossible to quantify in Barotrauma)

It is true that it was the case with WW2 submarines: the brushed DC dynamos of the German Type VII submarine, when cruising at her slowest possible speed of 60RPM / 1.5 knots, converted only 30% of the power coming from the battery as torque to the shaft. The remaining 70% was wasted mainly as heat in the main poles of the DC motor and also used for cooling. Of course this improved once speed increased: above 2.5 knots their efficiency was >70% in the remainder of the power band.

But this was 90 years ago, and Barotrauma is in the future. Compare the WW2 numbers to the efficiency curve of the prime mover aboard a modern conventional submarine. It has a flat band of >90% efficiency in most cases, even in the slow region where I assume is the ultra-silent running mode. This is closer to what I was thinking about when opening the ticket.
![image](https://github.com/Regalis11/Barotrauma/assets/68995233/61ab45b4-ce2b-407f-8e4e-accccfe091a9)
Source: Siemens (Public brochure of the Synavy Permasyn system)

I've made a chart to compare the efficiency curves of the Type VII WW2 sub, overlaid the brochure data (apologies for the crude drawing) and the Barotrauma curve in the current version.
![image](https://github.com/Regalis11/Barotrauma/assets/68995233/105f2a42-9e7d-4fc6-bd58-98bac58e35a3)


All of this has a huge effect on gameplay: the large difference means the engine draws so much more power than it should, that it makes little sense to attempt to conserve fuel by cruising slower.
I am aware that in the current version fuel is abundant however I hope it will become more scarce. Then this change will be even more important.

**Comment by Regalis11** [comment]
Thanks you for the explanation, I think I get it now. To me this doesn't seem like a bug though, but rather a request to make the efficiency curve more realistic, so I think discussions would be a more appropriate place for this.

**Reply by mygamingaccount** [reply]
Speed vs fuel consumption being something to consider could be additional content for captains, if the current steady flood of fuel rods were to be changed that is.
I've opened a PR for your consideration that implements these changes. Testing confirms the expected numbers: a 2x increase in speed requires an 8x increase in power, consistent with submarine hydrodynamics.

#12334

**Comment by G-M-twostay** [comment]
so with your change does it mean the sub will have more power/speed efficiency when at higher speed?

**Reply by mygamingaccount** [reply]
Power usage and thrust output are the same at maximum speed because $1^2 = 1$. The engine will consume less power at less than maximum speed - proportional with the power you get from $P = F * V$.
2x overvoltage (double power) will lead to a speed increase of 26% instead of the current 41%, as increasing speed by 41% would  need almost 3x more power, not 2x
